File: ros2_ws/src/navigation/mapping/mapping/k_means.py
#
# MOBILE ROBOTS - FI-UNAM, 2027-1
# CLUSTERING BY K-MEANS
#
# Instructions:
# 
#

import rclpy
from rclpy.node import Node
from rclpy.duration import Duration
from nav_msgs.msg import OccupancyGrid
from nav_msgs.srv import GetMap
from visualization_msgs.msg import Marker
from geometry_msgs.msg import Point
import numpy

FULL_NAME = "FULL NAME"

class KMeansNode(Node):
    def generate_random_centroids(self, K, min_x, min_y, max_x, max_y, static_map):
        centroids = numpy.zeros((K,2))
        for i in range(K):
            in_free_space = False
            while not in_free_space:
                x = (max_x - min_x) * numpy.random.rand() + min_x
                y = (max_y - min_y) * numpy.random.rand() + min_y
                c = int((x - static_map.info.origin.position.x)/static_map.info.resolution)
                r = int((y - static_map.info.origin.position.y)/static_map.info.resolution)
                idx = r*static_map.info.width + c
                in_free_space = static_map.data[idx] < 40 and static_map.data[idx] >= 0
            centroids[i,0] = x
            centroids[i,1] = y
        return centroids

    def recalculate_centroids(self, centroids, P):
        clusters = numpy.zeros(centroids.shape)
        counters = numpy.zeros((len(centroids), 1))
        new_centroids = numpy.zeros(centroids.shape)
        #
        # TODO:
        # Recalculate centroids
        #
        return new_centroids
            

    def k_means(self, P, K, static_map, tol):
        self.get_logger().info("Clustering " + str(len(P)) + "points with " + str(K) + " centroids")
        
        centroids = self.generate_random_centroids(K, -5, -5, 5, 5, static_map)
        self.pub_centroids.publish(self.get_centroids_marker(centroids))

        #
        # TODO:
        # Write the code to cluster P by K-means
        # P are the set of points to be clustered, K is the number of centroids,
        # static_map is the occupancy grid used to generate the first random centroids in the free space
        # tol is the tolerance to consider that the centroids have converged
        #
        max_distance = float("inf")
        while max_distance > tol:
            self.get_logger().info("Recalculating centroids")
            new_centroids = self.recalculate_centroids(centroids, P)
            max_distance = numpy.max([numpy.linalg.norm(centroids[i] - new_centroids[i]) for i in range(len(centroids))])
            centroids = new_centroids
            self.pub_centroids.publish(self.get_centroids_marker(centroids))
        self.get_logger().info("Converged K centroids after")
    
    def __init__(self):
        super().__init__("k_means_node")
        self.get_logger().info("INITIALIZING K-MEANS NODE - " + FULL_NAME)
        self.clt_inflated_map = self.create_client(GetMap, '/get_inflated_map')
        self.get_logger().info("Waiting for inflated map service...")
        while not self.clt_inflated_map.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Waiting for inflated map service...')
        self.get_logger().info("Inflated map service is now available...")
        future = self.clt_inflated_map.call_async(GetMap.Request())
        rclpy.spin_until_future_complete(self, future)
        response = future.result()
        self.static_map = response.map
        self.get_logger().info("Got inflated map.")
        self.declare_parameter('K', 3)
        self.declare_parameter('tol', 0.2)
        self.pub_centroids = self.create_publisher(Marker, '/mapping/k_means_centroids', 10)

    def get_centroids_marker(self, centroids):
        mrk = Marker()
        mrk.header.frame_id = "map"
        mrk.header.stamp = self.get_clock().now().to_msg()
        mrk.ns = "mapping"
        mrk.lifetime.sec=10000
        mrk.id = 0
        mrk.type   = Marker.SPHERE_LIST
        mrk.action = Marker.ADD
        mrk.scale.x, mrk.scale.y, mrk.scale.z = 0.2, 0.2, 0.2
        mrk.color.r, mrk.color.a = 1.0, 1.0
        mrk.points = [Point(x=c[0], y=c[1], z=0.1) for c in centroids]
        return mrk

    def get_cartesian_free_points(self, static_map):
        P = []
        for r in range(static_map.info.height):
            for c in range(static_map.info.width):
                idx = r*static_map.info.width + c
                if static_map.data[idx] >= 0 and static_map.data[idx] < 40:
                    x = c*static_map.info.resolution + static_map.info.origin.position.x
                    y = r*static_map.info.resolution + static_map.info.origin.position.y
                    P.append([x,y])
        return numpy.asarray(P)
        

    def spin(self):
        K = self.get_parameter('K').get_parameter_value().integer_value
        tol = self.get_parameter('tol').get_parameter_value().double_value
        P = self.get_cartesian_free_points(self.static_map)
        self.k_means(P, K, self.static_map, tol)
        while rclpy.ok():
            rclpy.spin_once(self, timeout_sec=0)
            self.get_clock().sleep_for(Duration(seconds=0.1))

def main(args=None):
    rclpy.init(args=args)
    k_means_node = KMeansNode()
    k_means_node.spin()
    k_means_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros2_ws/src/navigation/mapping_cpp/src/k_means.cpp
/*
 * MOBILE ROBOTS - FI-UNAM, 2027-1
 * CLUSTERING BY K-MEANS
 * Instructions:
 * Write the code necessary to implement the K-means algorithm to cluster a map
 * given an occupancy grid and a number of centroids K
 * Modify only the sections with the TODO comment
 */

#include <cstdio>
#include <limits>
#include <Eigen/Dense>
#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/srv/get_map.hpp"
#include "random_numbers/random_numbers.h"
#include "visualization_msgs/msg/marker.hpp"

#define FULL_NAME "LEONARDO ALEJANDRO GARCIA LERMA"

class KMeansNode : public rclcpp::Node
{
public:
    KMeansNode():Node("k_means"){
	RCLCPP_INFO(this->get_logger(), "INITIALIZING K-MEANS NODE - %s", FULL_NAME);
	K = this->declare_parameter<int>("K", 3);
	tol = this->declare_parameter<double>("tol", 0.2);
	clt_inflated_map = this->create_client<nav_msgs::srv::GetMap>("/get_inflated_map");
	pub_marker = this->create_publisher<visualization_msgs::msg::Marker>("/mapping/k_means_centroids", 1);
    }

    std::vector<Eigen::Vector2d> get_cartesian_free_points(nav_msgs::msg::OccupancyGrid& map){
	std::vector<Eigen::Vector2d> P;
	for(size_t i=0; i < map.data.size(); i++){
	    if(map.data[i] != 0)
		continue;
	    int row = i / map.info.width;
	    int col = i % map.info.width;
	    double x = col * map.info.resolution + map.info.origin.position.x;
	    double y = row * map.info.resolution + map.info.origin.position.y;
	    Eigen::Vector2d p(x,y);
	    P.push_back(p);
	}
	RCLCPP_INFO(this->get_logger(), "Got map with %lu points in free space", P.size());
	return P;
    }

    std::vector<Eigen::Vector2d> generate_random_centroids(int K, double min_x, double min_y, double max_x, double max_y,
							   nav_msgs::msg::OccupancyGrid& map){
	std::vector<Eigen::Vector2d> C;
	random_numbers::RandomNumberGenerator rnd;
	for(int i=0; i < K; i++){
	    bool in_free_space = false;
	    double x,y;
	    while(!in_free_space){
		x = rnd.uniformReal(min_x, max_x);
		y = rnd.uniformReal(min_y, max_y);
		int row = (int)((y - map.info.origin.position.y)/map.info.resolution);
		int col = (int)((x - map.info.origin.position.x)/map.info.resolution);
		in_free_space = map.data[row*map.info.width + col] == 0;
	    }
	    Eigen::Vector2d c(x,y);
	    C.push_back(c);
	}
	RCLCPP_INFO(this->get_logger(), "Generated %lu random centroids", C.size());
	return C;
    }

    std::vector<Eigen::Vector2d> recalculate_centroids(std::vector<Eigen::Vector2d>& centroids,
						       std::vector<Eigen::Vector2d>& points){
	RCLCPP_INFO(this->get_logger(), "Recalculating centroids");
	std::vector<Eigen::Vector2d> new_centroids(centroids.size());	    
	std::vector<int> counters(centroids.size());
	/*
	 * TODO:
	 * Implement the steps to recalculate centroids.
	 * Use as reference the python version of this algorithm.
	 * Use the declared variables and the Eigen library
	 */
	while (max_distance > tol) {
    RCLCPP_INFO(this->get_logger(), "Recalculating centroids");

    auto new_centroids = recalculate_centroids(centroids, P);

    // Calcular la distancia máxima entre los centroides viejos y nuevos
    max_distance = 0.0;
    for (size_t i = 0; i < centroids.size(); ++i) {
        double dist = std::hypot(centroids[i].x - new_centroids[i].x,
                                 centroids[i].y - new_centroids[i].y);
        if (dist > max_distance) {
            max_distance = dist;
        }
    }

    centroids = new_centroids;
    iterations += 1;
    pub_centroids_->publish(get_centroids_marker(centroids));
}
	/*
	 * END OF TODO
	 */
	return new_centroids;
    }

    visualization_msgs::msg::Marker get_centroids_marker(std::vector<Eigen::Vector2d>& centroids){
	visualization_msgs::msg::Marker mrk;
	mrk.header.frame_id = "map";
        mrk.header.stamp = this->get_clock()->now();
        mrk.ns = "mapping";
	mrk.lifetime.sec=10000;
        mrk.id = 0;
        mrk.type   = visualization_msgs::msg::Marker::SPHERE_LIST;
        mrk.action = visualization_msgs::msg::Marker::ADD;
	mrk.scale.x = 0.2;
	mrk.scale.y = 0.2;
	mrk.scale.z = 0.2;
	mrk.color.r = 1.0;
	mrk.color.a = 1.0;
	for(size_t i=0; i < centroids.size(); i++){
	    geometry_msgs::msg::Point p;
	    p.x = centroids[i][0];
	    p.y = centroids[i][1];
	    mrk.points.push_back(p);
	}
	return mrk;
    }

    void spin()
    {
	/*
	 * Waiting for inflated map service to be available
	 */
	RCLCPP_INFO(this->get_logger(), "Waiting for inflated map");
	while(!this->clt_inflated_map->wait_for_service(std::chrono::seconds(1))){
	    if(!rclcpp::ok()){
		RCLCPP_INFO(this->get_logger(), "Interrupted while waiting for the inflated map service. Exiting.");
		return;
	    }
	    RCLCPP_INFO(this->get_logger(), "Inflated map service not available, waiting again...");
	}
	/*
	 * Getting inflated map
	 */
	nav_msgs::msg::OccupancyGrid inflated_map;
	RCLCPP_INFO(this->get_logger(), "Inflated map service is now available. Getting inflated map.");
	nav_msgs::srv::GetMap::Request::SharedPtr req = std::make_shared<nav_msgs::srv::GetMap::Request>();
	rclcpp::Client<nav_msgs::srv::GetMap>::FutureAndRequestId result = this->clt_inflated_map->async_send_request(req);
	if(rclcpp::spin_until_future_complete(this->get_node_base_interface(), result) == rclcpp::FutureReturnCode::SUCCESS){
	    inflated_map = result.get()->map;
	    RCLCPP_INFO(this->get_logger(), "Got map with size %d x %d", inflated_map.info.width, inflated_map.info.height);
	}else{
	    RCLCPP_INFO(this->get_logger(), "Cannot get inflated map. Aborting node. ");
	    return;
	}
	
	std::vector<Eigen::Vector2d> P = get_cartesian_free_points(inflated_map);
	std::vector<Eigen::Vector2d> centroids = generate_random_centroids(this->K, -5, -5, 10, 5,  inflated_map);
	visualization_msgs::msg::Marker mrk = get_centroids_marker(centroids);
	pub_marker->publish(mrk);

	double max_dist = std::numeric_limits<double>::max();
	while(rclcpp::ok() && max_dist > this->tol){
	    pub_marker->publish(mrk);
	    rclcpp::spin_some(this->get_node_base_interface());
	    rclcpp::sleep_for(std::chrono::milliseconds(100));
	    /*
	     * TODO:
	     * Recalculate centroids
	     * Get the maximum distance between each centroids and is corresponding new centroid
	     * Use as reference the python version of this algorithm.
	     * Use the declared variables and the Eigen library
	     */
	    for (const auto& p : P) {
    int best_idx = 0;
    double min_dist = std::numeric_limits<double>::max();

    // Encontrar el centroide más cercano
    for (size_t c = 0; c < centroids.size(); ++c) {
        double dist = std::hypot(p.x - centroids[c].x, p.y - centroids[c].y);
        if (dist < min_dist) {
            min_dist = dist;
            best_idx = c;
        }
    }

    // Sumar las coordenadas del punto al cluster correspondiente
    clusters[best_idx].x += p.x;
    clusters[best_idx].y += p.y;
    counters[best_idx] += 1;
}

// Calcular el promedio para obtener los nuevos centroides
for (size_t i = 0; i < centroids.size(); ++i) {
    if (counters[i] > 0) {
        new_centroids[i].x = clusters[i].x / counters[i];
        new_centroids[i].y = clusters[i].y / counters[i];
    }
}
	    /*
	     * END OF TODO
	     */
	    RCLCPP_INFO(this->get_logger(), "Max change in centroids: %lf", max_dist);
	    mrk = get_centroids_marker(centroids);
	}
    }

private:
    int K;
    double tol;
    rclcpp::Client<nav_msgs::srv::GetMap>::SharedPtr clt_inflated_map{nullptr};
    rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr pub_marker{nullptr};
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto n = std::make_shared<KMeansNode>();
  n->spin();
  rclcpp::shutdown();
  return 0;
}

